File: packages/tale_shared/src/tale_shared/config/providers.py
"""Provider configuration reader for file-based LLM provider config."""

import json
import logging
import os
import subprocess
from dataclasses import dataclass, field
from pathlib import Path

from tale_shared.config.org_slug import validate_org_slug
from tale_shared.utils.sops import decrypt_secrets_file

logger = logging.getLogger(__name__)

DEFAULT_CONFIG_DIR = "/app/data"


@dataclass
class ModelConfig:
    """A single model definition within a provider."""

    id: str
    display_name: str
    tags: list[str]
    description: str = ""
    dimensions: int | None = None
    # Optional per-model env-var name holding the API key (issue #1711).
    secrets_env: str | None = None


@dataclass
class ProviderConfig:
    """A provider loaded from a JSON file, with optional decrypted secrets."""

    name: str
    display_name: str
    base_url: str
    models: list[ModelConfig] = field(default_factory=list)
    description: str = ""
    supports_structured_outputs: bool = False
    api_key: str | None = None
    defaults: dict[str, str] = field(default_factory=dict)
    # Optional provider-level env-var name holding the API key (issue #1711).
    secrets_env: str | None = None


# Deployment allowlist (empty = locked) gating which env-var names may source a
# provider API key. Mirrors the platform's `TALE_PROVIDER_SECRET_ENV_ALLOWLIST`
# so the platform and the Python services resolve the same effective key. See
# `services/platform/convex/providers/secret_resolver.ts`.
_ALLOWLIST_ENV = "TALE_PROVIDER_SECRET_ENV_ALLOWLIST"


def _env_secret(name: str | None) -> str | None:
    """Resolve an env-var name to its trimmed value, honoring the allowlist.

    Returns None when the name is missing, not allowlisted, or the env var is
    empty/whitespace. Trailing-newline normalization (a common Vault/k8s
    injection footgun) is applied to env values here.
    """
    if not name:
        return None
    raw = os.environ.get(_ALLOWLIST_ENV, "")
    allowlist = {n.strip() for n in raw.split(",") if n.strip()}
    if name not in allowlist:
        logger.warning(
            "secretsEnv %r is not in %s (empty allowlist disables the env key "
            "source) — falling back to the secrets file",
            name,
            _ALLOWLIST_ENV,
        )
        return None
    value = os.environ.get(name, "").strip()
    if not value:
        logger.warning(
            "secretsEnv %r is set but the env var is empty/unset — falling back to the secrets file",
            name,
        )
        return None
    return value


def _resolve_api_key(provider: "ProviderConfig", model: "ModelConfig") -> str:
    """Resolve the effective API key for a model (issue #1711).

    Order: model env -> provider env -> provider file key. The Python loader has
    no per-model file-key tier (only `apiKey`), so this is 3-tier. Returns ""
    (never None) when nothing resolves, preserving the previous
    ``provider.api_key or ""`` contract and the config-equality used by the
    rag/crawler per-org client caches.
    """
    for name in (model.secrets_env, provider.secrets_env):
        value = _env_secret(name)
        if value:
            return value
    return provider.api_key or ""


def load_providers(
    org_slug: str,
    config_dir: str | None = None,
) -> list[ProviderConfig]:
    """Read all provider JSON files from {config_dir}/{org_slug}/providers/.

    Under the org-first config layout, each org owns its own provider
    catalog at `<root>/<org_slug>/providers/`. `org_slug` is required —
    pinning RAG/crawler globally to the `default` org's providers would
    quietly serve the wrong models to other orgs.

    Reads *.json (excluding *.secrets.json) and decrypts matching
    *.secrets.json files via SOPS.
    """
    # Defense in depth: the FastAPI deps in RAG / crawler already gate
    # `X-Tale-Org` against the same regex at request boundary, but
    # internal callers in long-running tasks (e.g. crawler scheduler,
    # vision hot paths) reach this function with a slug taken from
    # `get_active_org()` or other module state. A `.` or `/etc` slipping
    # in here would silently route to `<base>/providers` or `/etc/
    # providers` via Path()'s "absolute resets" / "dot is a no-op"
    # semantics — exactly the legacy-flat-layout class the org-first
    # refactor exists to retire. Validate at the boundary, not just on
    # the way in. (Round-2 P1-30.)
    validate_org_slug(org_slug)

    shared_config = os.environ.get("TALE_PLATFORM_SHARED_CONFIG_DIR")
    if shared_config:
        base = Path(shared_config)
    else:
        base = Path(config_dir or os.environ.get("TALE_CONFIG_DIR") or os.environ.get("CONFIG_DIR", DEFAULT_CONFIG_DIR))
    providers_dir = base / org_slug / "providers"

    if not providers_dir.is_dir():
        logger.warning(
            "Providers directory not found for org '%s': %s",
            org_slug,
            providers_dir,
        )
        return []

    providers: list[ProviderConfig] = []

    for json_file in sorted(providers_dir.glob("*.json")):
        if json_file.name.endswith(".secrets.json"):
            continue

        try:
            with open(json_file) as f:
                data = json.load(f)
        except (json.JSONDecodeError, OSError) as exc:
            logger.error("Failed to read provider file %s: %s", json_file, exc)
            continue

        provider_name = json_file.stem

        # Load secrets if present
        api_key: str | None = None
        secrets_file = json_file.with_suffix("").with_suffix(".secrets.json")
        if secrets_file.exists():
            try:
                secrets = decrypt_secrets_file(secrets_file)
                api_key = secrets.get("apiKey")
            except (RuntimeError, OSError, subprocess.TimeoutExpired) as exc:
                logger.error("Failed to decrypt secrets for %s: %s", provider_name, exc)

        models = []
        for m in data.get("models", []):
            models.append(
                ModelConfig(
                    id=m["id"],
                    display_name=m.get("displayName", m["id"]),
                    tags=m.get("tags", []),
                    description=m.get("description", ""),
                    dimensions=m.get("dimensions"),
                    secrets_env=m.get("secretsEnv"),
                )
            )

        # Read defaults map, migrating legacy per-model default if needed
        defaults: dict[str, str] = {}
        if "defaults" in data and isinstance(data["defaults"], dict):
            defaults = {k: v for k, v in data["defaults"].items() if isinstance(k, str) and isinstance(v, str)}
        else:
            # Migrate legacy format: model-level default: true
            for m in data.get("models", []):
                if m.get("default") is True:
                    for tag in m.get("tags", []):
                        if tag not in defaults:
                            defaults[tag] = m["id"]

        providers.append(
            ProviderConfig(
                name=provider_name,
                display_name=data.get("displayName", provider_name),
                base_url=data.get("baseUrl", ""),
                models=models,
                description=data.get("description", ""),
                supports_structured_outputs=data.get("supportsStructuredOutputs", False),
                api_key=api_key,
                defaults=defaults,
                secrets_env=data.get("secretsEnv"),
            )
        )

    return providers


def _find_model(
    providers: list[ProviderConfig], tag: str, *, prefer_default: bool = False
) -> tuple[ProviderConfig, ModelConfig] | None:
    """Find a model by tag across all providers.

    If prefer_default is True, check the provider-level defaults map first,
    falling back to the first model with the tag.
    """
    if prefer_default:
        for provider in providers:
            default_model_id = provider.defaults.get(tag)
            if default_model_id:
                for model in provider.models:
                    if model.id == default_model_id:
                        return (provider, model)

    for provider in providers:
        for model in provider.models:
            if tag in model.tags:
                return (provider, model)

    return None


def get_chat_model(
    org_slug: str,
    config_dir: str | None = None,
) -> tuple[str, str, str]:
    """Return (base_url, api_key, model_id) for the org's default chat model.

    Finds the first model marked default that has a "chat" tag,
    or falls back to the first model with a "chat" tag.
    """
    providers = load_providers(org_slug, config_dir)
    match = _find_model(providers, "chat", prefer_default=True)
    if match is None:
        raise ValueError(f"No chat model found in provider configuration files for org '{org_slug}'.")

    provider, model = match
    api_key = _resolve_api_key(provider, model)
    return (provider.base_url, api_key, model.id)


def get_embedding_model(
    org_slug: str,
    config_dir: str | None = None,
) -> tuple[str, str, str, int]:
    """Return (base_url, api_key, model_id, dimensions) for the org's embedding model."""
    providers = load_providers(org_slug, config_dir)
    match = _find_model(providers, "embedding", prefer_default=True)
    if match is None:
        raise ValueError(f"No embedding model found in provider configuration files for org '{org_slug}'.")

    provider, model = match
    api_key = _resolve_api_key(provider, model)
    dims = model.dimensions
    if dims is None:
        raise ValueError(
            f"Embedding model {model.id} does not specify dimensions. Add a 'dimensions' field to the model definition."
        )
    return (provider.base_url, api_key, model.id, dims)


def get_vision_model(
    org_slug: str,
    config_dir: str | None = None,
) -> tuple[str, str, str]:
    """Return (base_url, api_key, model_id) for the org's vision model."""
    providers = load_providers(org_slug, config_dir)
    match = _find_model(providers, "vision", prefer_default=True)
    if match is None:
        raise ValueError(f"No vision model found in provider configuration files for org '{org_slug}'.")

    provider, model = match
    api_key = _resolve_api_key(provider, model)
    return (provider.base_url, api_key, model.id)

File: packages/shared/src/config/providers.ts
/**
 * Provider configuration reader for file-based LLM provider config.
 *
 * Under the org-first config layout each org owns its own provider catalog at
 * `<root>/<org_slug>/providers/`. `orgSlug` is required — pinning RAG/crawler
 * globally to the `default` org's providers would quietly serve the wrong
 * models to other orgs.
 */

import { existsSync, readdirSync, readFileSync, statSync } from 'node:fs';
import path from 'node:path';

import { z } from 'zod';

import { decryptSecretsFile } from '../utils/sops';
import { validateOrgSlug } from './org-slug';

const DEFAULT_CONFIG_DIR = '/app/data';

/**
 * Reserved prefix every env-var name sourcing a provider API key must carry
 * (issue #1711). Fail-closed: any name outside the prefix is rejected. Mirrors
 * `SECRETS_ENV_PREFIX` in `services/platform/lib/shared/schemas/providers.ts`
 * and the runtime gate in `services/platform/convex/providers/secret_resolver.ts`.
 */
const SECRETS_ENV_PREFIX = 'TALE_PROVIDER_KEY_';

export interface ModelConfig {
  id: string;
  displayName: string;
  tags: string[];
  description: string;
  dimensions: number | null;
  secretsEnv: string | null;
}

export interface ProviderConfig {
  name: string;
  displayName: string;
  baseUrl: string;
  models: ModelConfig[];
  description: string;
  supportsStructuredOutputs: boolean;
  apiKey: string | null;
  defaults: Record<string, string>;
  secretsEnv: string | null;
}

export type ChatModel = { baseUrl: string; apiKey: string; modelId: string };
export type EmbeddingModel = ChatModel & { dimensions: number };
export type VisionModel = ChatModel;

const stringRecord = z.record(z.string(), z.string());

const rawModelSchema = z.looseObject({
  id: z.string(),
  displayName: z.string().optional(),
  tags: z.array(z.string()).optional(),
  description: z.string().optional(),
  dimensions: z.number().optional(),
  secretsEnv: z.unknown().optional(),
  default: z.unknown().optional(),
});

const rawProviderSchema = z.looseObject({
  displayName: z.string().optional(),
  baseUrl: z.string().optional(),
  description: z.string().optional(),
  supportsStructuredOutputs: z.boolean().optional(),
  secretsEnv: z.unknown().optional(),
  defaults: z.unknown().optional(),
  models: z.array(rawModelSchema).optional(),
});

/**
 * Resolve an env-var name to its trimmed value, honoring the reserved prefix.
 *
 * Returns null when the name is missing, not a string, not prefixed, or the
 * env var is empty/whitespace. Trailing-newline normalization (a common
 * Vault/k8s injection footgun) is applied to env values here. A hand-edited
 * non-string `secretsEnv` degrades to the file fallback rather than throwing.
 */
function envSecret(name: unknown): string | null {
  if (typeof name !== 'string' || name.length === 0) {
    return null;
  }
  if (!name.startsWith(SECRETS_ENV_PREFIX)) {
    console.warn(
      `secretsEnv ${JSON.stringify(name)} does not start with the reserved prefix ` +
        `${SECRETS_ENV_PREFIX} — falling back to the secrets file`,
    );
    return null;
  }
  const value = (process.env[name] ?? '').trim();
  if (!value) {
    console.warn(
      `secretsEnv ${JSON.stringify(name)} is set but the env var is empty/unset — ` +
        `falling back to the secrets file`,
    );
    return null;
  }
  return value;
}

/**
 * Resolve the effective API key for a model (issue #1711).
 *
 * Order: model env -> provider env -> provider file key. Returns "" (never
 * null) when nothing resolves, preserving the config-equality used by the
 * rag/crawler per-org client caches.
 */
function resolveApiKey(provider: ProviderConfig, model: ModelConfig): string {
  for (const name of [model.secretsEnv, provider.secretsEnv]) {
    const value = envSecret(name);
    if (value) {
      return value;
    }
  }
  return provider.apiKey ?? '';
}

function resolveBaseDir(configDir: string | null | undefined): string {
  const sharedConfig = process.env.TALE_PLATFORM_SHARED_CONFIG_DIR;
  if (sharedConfig) {
    return sharedConfig;
  }
  return (
    configDir ??
    process.env.TALE_CONFIG_DIR ??
    process.env.CONFIG_DIR ??
    DEFAULT_CONFIG_DIR
  );
}

/**
 * Read all provider JSON files from `{configDir}/{orgSlug}/providers/`.
 *
 * Reads `*.json` (excluding `*.secrets.json`) and decrypts matching
 * `*.secrets.json` files via SOPS.
 */
export function loadProviders(
  orgSlug: string,
  configDir?: string | null,
): ProviderConfig[] {
  validateOrgSlug(orgSlug);

  const base = resolveBaseDir(configDir);
  const providersDir = path.join(base, orgSlug, 'providers');

  if (!existsSync(providersDir) || !statSync(providersDir).isDirectory()) {
    console.warn(
      `Providers directory not found for org '${orgSlug}': ${providersDir}`,
    );
    return [];
  }

  const jsonFiles = readdirSync(providersDir)
    .filter((name) => name.endsWith('.json') && !name.endsWith('.secrets.json'))
    .sort();

  const providers: ProviderConfig[] = [];

  for (const fileName of jsonFiles) {
    const jsonFile = path.join(providersDir, fileName);

    let raw: z.infer<typeof rawProviderSchema>;
    try {
      raw = rawProviderSchema.parse(
        JSON.parse(readFileSync(jsonFile, 'utf-8')),
      );
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      console.error(`Failed to read provider file ${jsonFile}: ${message}`);
      continue;
    }

    const providerName = path.basename(fileName, '.json');

    let apiKey: string | null = null;
    const secretsFile = path.join(providersDir, `${providerName}.secrets.json`);
    if (existsSync(secretsFile)) {
      try {
        const secrets = decryptSecretsFile(secretsFile);
        const candidate = secrets.apiKey;
        apiKey = typeof candidate === 'string' ? candidate : null;
      } catch (err) {
        const message = err instanceof Error ? err.message : String(err);
        console.error(
          `Failed to decrypt secrets for ${providerName}: ${message}`,
        );
      }
    }

    const rawModels = raw.models ?? [];
    const models: ModelConfig[] = rawModels.map((m) => ({
      id: m.id,
      displayName: m.displayName ?? m.id,
      tags: m.tags ?? [],
      description: m.description ?? '',
      dimensions: m.dimensions ?? null,
      secretsEnv: typeof m.secretsEnv === 'string' ? m.secretsEnv : null,
    }));

    let defaults: Record<string, string> = {};
    const rawDefaults = stringRecord.safeParse(raw.defaults);
    if (rawDefaults.success) {
      defaults = rawDefaults.data;
    } else {
      // Migrate legacy format: model-level `default: true`.
      for (const m of rawModels) {
        if (m.default === true) {
          for (const tag of m.tags ?? []) {
            if (!(tag in defaults)) {
              defaults[tag] = m.id;
            }
          }
        }
      }
    }

    providers.push({
      name: providerName,
      displayName: raw.displayName ?? providerName,
      baseUrl: raw.baseUrl ?? '',
      models,
      description: raw.description ?? '',
      supportsStructuredOutputs: raw.supportsStructuredOutputs ?? false,
      apiKey,
      defaults,
      secretsEnv: typeof raw.secretsEnv === 'string' ? raw.secretsEnv : null,
    });
  }

  return providers;
}

/**
 * Find a model by tag across all providers. If `preferDefault` is true, check
 * the provider-level defaults map first, falling back to the first model with
 * the tag.
 */
function findModel(
  providers: ProviderConfig[],
  tag: string,
  preferDefault: boolean,
): { provider: ProviderConfig; model: ModelConfig } | null {
  if (preferDefault) {
    for (const provider of providers) {
      const defaultModelId = provider.defaults[tag];
      if (defaultModelId) {
        for (const model of provider.models) {
          if (model.id === defaultModelId) {
            return { provider, model };
          }
        }
      }
    }
  }

  for (const provider of providers) {
    for (const model of provider.models) {
      if (model.tags.includes(tag)) {
        return { provider, model };
      }
    }
  }

  return null;
}

/** Return the chat model config for the org's default chat model. */
export function getChatModel(
  orgSlug: string,
  configDir?: string | null,
): ChatModel {
  const providers = loadProviders(orgSlug, configDir);
  const match = findModel(providers, 'chat', true);
  if (match === null) {
    throw new Error(
      `No chat model found in provider configuration files for org '${orgSlug}'.`,
    );
  }
  return {
    baseUrl: match.provider.baseUrl,
    apiKey: resolveApiKey(match.provider, match.model),
    modelId: match.model.id,
  };
}

/** Return the embedding model config (incl. dimensions) for the org. */
export function getEmbeddingModel(
  orgSlug: string,
  configDir?: string | null,
): EmbeddingModel {
  const providers = loadProviders(orgSlug, configDir);
  const match = findModel(providers, 'embedding', true);
  if (match === null) {
    throw new Error(
      `No embedding model found in provider configuration files for org '${orgSlug}'.`,
    );
  }
  const { dimensions } = match.model;
  if (dimensions === null) {
    throw new Error(
      `Embedding model ${match.model.id} does not specify dimensions. Add a ` +
        `'dimensions' field to the model definition.`,
    );
  }
  return {
    baseUrl: match.provider.baseUrl,
    apiKey: resolveApiKey(match.provider, match.model),
    modelId: match.model.id,
    dimensions,
  };
}

/** Return the vision model config for the org. */
export function getVisionModel(
  orgSlug: string,
  configDir?: string | null,
): VisionModel {
  const providers = loadProviders(orgSlug, configDir);
  const match = findModel(providers, 'vision', true);
  if (match === null) {
    throw new Error(
      `No vision model found in provider configuration files for org '${orgSlug}'.`,
    );
  }
  return {
    baseUrl: match.provider.baseUrl,
    apiKey: resolveApiKey(match.provider, match.model),
    modelId: match.model.id,
  };
}
